### PR TITLE
Add Feature Policy

### PR DESCRIPTION
**Pull request by meyskens**
Document some policies for features inside the cert-manager codebase

Feedback very welcome!

/cc @munnerz

### CODE REVIEW

File: content/en/docs/contributing/policy.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+linkTitle: "Feature Policy"
+weight: 70
+type: "docs"
+---
+
+We are open to feature requests and PRs implementing those. If you plan on contributing a new feature into cert-manager we recommend creating an issue first for it to be discussed with the cert-manager maintainers. Another possibility is bringing it up in a community meeting discussion for an open discussion with the maintainers and community on the implementation.
+
+# Features we will not allow
+
+Certain features have ben requested before but are not able to being implemented properly or not able to being implemented without breaking the security model.
```
Comment (deads2k) [inline content/en/docs/contributing/policy.md:12]: a few typos

Certain features have been requested before, but have not been implemented properly or not able to be implemented without breaking the security model.

File: content/en/docs/contributing/policy.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Below is a list of a few of these:
+
+## Vendoring Kubernetes related APIs outside of the `k8s.io/` namespace
+
+Vendoring Kubernetes related APIs like OpenShift specific classes is not recommend as it vendors the `k8s.io/apimachinery` package itself could potentially conflict with cert-manager's instance. It could also cause a conflict with different Kubernetes client versions being used.
```
Comment (deads2k) [inline content/en/docs/contributing/policy.md:18]: I would rephrase slightly to describe the generic problem with specific examples.

Vendoring project APIs that also vendor `k8s.io/apimachinery`, such as OpenShift, Contour, or Velero, is not recommend because the kubernetes dependency is likely to conflict with cert-manager's instance.

File: content/en/docs/contributing/policy.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+Helm deployments are namespaced by design and do not support a dependency model that would support cert-manager being operated on a cluster scope nor preventing cert-manager from being installed twice (with conflicting versions) or being upgraded independently from the application.
+
+The cert-manager installation creates cluster scoped resources like admission webhooks and custom resource definitions. cert-manager should be seen as part of your cluster and should be treated as such for being installed. A comparison to other Kubernetes components would be a LoadBalancer controller or a PV provisioner
+
+## Secret injection or copying
+
+cert-manager deals with very sensitive information (all TLS certificates for your services) and has cluster-level access to secret resources, therefore when designing features we need to consider all ways these can be abused to escalate privilege.
+Secret data is meant to be securely stored in the secret resources and have narrow scoped access privileges for unauthorized users. Therefore we will not allow any functionality that allows this data to be copied/injected into any non secret resource.
```
Comment (deads2k) [inline content/en/docs/contributing/policy.md:41]: into any resource other than a kubernetes secret (`/api/v1/namespaces/<ns>/secrets`).

File: content/en/docs/contributing/policy.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+The cert-manager installation creates cluster scoped resources like admission webhooks and custom resource definitions. cert-manager should be seen as part of your cluster and should be treated as such for being installed. A comparison to other Kubernetes components would be a LoadBalancer controller or a PV provisioner
+
+## Secret injection or copying
+
+cert-manager deals with very sensitive information (all TLS certificates for your services) and has cluster-level access to secret resources, therefore when designing features we need to consider all ways these can be abused to escalate privilege.
+Secret data is meant to be securely stored in the secret resources and have narrow scoped access privileges for unauthorized users. Therefore we will not allow any functionality that allows this data to be copied/injected into any non secret resource. 
+
+### cainjector
+
+The cainjector component is a special exception to this rule. This component is able to inject the `ca.crt` file into predefined fields on `ValidatingWebhookConfiguration`, `MutatingWebhookConfiguration`, and `CustomResourceDefinition` resources from Certificate resources.
```
Comment (deads2k) [inline content/en/docs/contributing/policy.md:45]: The cainjector component deals in non-sensitive information (CAs, not cert/key pairs).  This component is able to inject....

Comment (deads2k) [inline content/en/docs/contributing/policy.md:45]: If you're open to injecting into configmaps (I think this is probably reasonable), I would indicate that configmaps are anticipated to be the outward boundary/API instead of injecting into resources managed by other projects.

File: content/en/docs/contributing/policy.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+## Cross namespace resources
+
+Namespace boundaries in Kubernetes provide a barrier for access scopes. Apps or users can be limited to only access resources in a certain namespace. cert-manager is a controller that operates on cluster wide resources, while it may seems interesting to allow access to copy or write certificate data from one namespace to the other it will cause a bypass of this security model for all our users which is not intended and a security issue.
+If this behavior is intended for your use case there are other Kubernetes controllers that will do this for you however we want to suggest a certain caution when installing these. 
+
+## Sign certificates using the Kubernetes CA (used for your nodes)
+
+
+Kubernetes has a Certificate Signing Requests API, and a `kubectl certificates` command which allows you to approve certificate signing requests and have them signed by the certificate authority (CA) of the Kubernetes cluster.
```
Comment (deads2k) [inline content/en/docs/contributing/policy.md:58]: Have a look at the v1 API.  The trust for kubernetes clusters is actually split into three different signers as opposed to a "the certificate authority of the kubernetes cluster".  There isn't a single authority and CA bundles are not provided in SA token secrets.

Comment (meyskens) [reply content/en/docs/contributing/policy.md:58]: Need to dig deeper myself into this, this comes from our FAQ section. Proably needs a rewrite

File: content/en/docs/contributing/policy.md (hunk 4 above)
Comment (meyskens) [reply content/en/docs/contributing/policy.md:45]: Going to leave that out for now as there is a discussion going on about configmaps

**Review by wallrj — APPROVED** [review]
@meyskens shall we merge this?

I think it all makes sense and answers some FAQs so it'll be nice to be able to direct people here in future.

/hold
/lgtm

Unhold when you're happy for it to merge.

**Review by wallrj — APPROVED** [review]
There have been no further comments since my last review,
so I'll remove the hold  and merge this.

We can make further improvements / amendments in  followup PRs.

/unhold
